sql/src/main/java/io/crate/execution/engine/window/WindowBatchIterator.java

```diff
@@ -0,0 +1,133 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.engine.window;
+
+import io.crate.analyze.WindowDefinition;
+import io.crate.data.BatchIterator;
+import io.crate.data.CollectingRowConsumer;
+import io.crate.data.InMemoryBatchIterator;
+import io.crate.data.MappedForwardingBatchIterator;
+import io.crate.data.Row;
+import io.crate.execution.engine.aggregation.AggregateCollector;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+public class WindowBatchIterator extends MappedForwardingBatchIterator<Row, Row> {
+
+    private final BatchIterator<Row> source;
+    private final AggregateCollector collector;
+    private WindowDefinition windowDefinition;
+
+    private Row currentElement;
+
+    private int sourceRowsConsumed;
+    private int rowPosition;
+    private List<Row> peerRows = new ArrayList<>();
+    private boolean peersCollected = false;
+    private Row prevRow = null;
+
+    WindowBatchIterator(WindowDefinition windowDefinition, BatchIterator<Row> source, AggregateCollector collector) {
+        assert windowDefinition.partitions().size() == 0 : "Only empty OVER() is supported now";
+        assert windowDefinition.orderBy() == null : "Only empty OVER() is supported now";
+        assert windowDefinition.windowFrameDefinition() == null : "Only empty OVER() is supported now";
+
+        this.windowDefinition = windowDefinition;
+        this.source = source;
+        this.collector = collector;
+    }
+
+    private boolean arePeers(@Nullable Row prevRow, Row nextRow) {
+        // all rows are peers when orderBy is missing
+        return windowDefinition.orderBy() == null;
+    }
+
+    @Override
+    protected BatchIterator<Row> delegate() {
+        return source;
+    }
+
+    @Override
+    public Row currentElement() {
+        return currentElement;
+    }
+
+    @Override
+    public void moveToStart() {
+        super.moveToStart();
+        sourceRowsConsumed = 0;
+        rowPosition = 0;
+        peerRows = new ArrayList<>();
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (peersCollected && rowPosition < sourceRowsConsumed) {
+            rowPosition++;
+            return true;
+        }
+
+        while (source.moveNext()) {
+            sourceRowsConsumed++;
+            Row currentRow = source.currentElement();
+            if (arePeers(prevRow, currentRow)) {
+                peersCollected = false;
+                peerRows.add(currentRow);
```

Comment:
rows have to be materialized in order to safely store them across `moveNext` calls